File: leetcode/HappyNumber/isHappy.java
public class Solution {
  public boolean isHappy(int n) {
    HashSet<Integer> set = new HashSet<Integer>();
    return helper(n, set);
  }
  private boolean helper(int n, HashSet<Integer> set) {
    if (set.contains(n)) return false;
    set.add(n);
    int sum = 0;
    while (n > 0) {
      int digit = n % 10;
      sum += digit * digit;
      n /= 10;
    }
    if (sum == 1) return true;
    return helper(sum, set);
  }
}

File: leetcode/HappyNumber/isHappy.cpp
class Solution {
public:
  bool isHappy(int n) {
    unordered_set<int> map;
    return helper(n, &map);
  }
  bool helper(int n, unordered_set<int> *map) {
    if (map->find(n) != map->end()) return false;
    map->insert(n);
    if (n == 1) return true;
    int ans = 0;
    while (n) {
      int digit = n % 10;
      ans += digit * digit;
      n /= 10;
    }
    return helper(ans, map);
  }
};
